### PR TITLE
Nitmd benches

### PR DESCRIPTION
**Pull request by Morriar**
Copied from README:

# Bench Markdown

Benches markdown parsers.

## Usage

        ./bench_markdown.sh all

## Engines

* nitmd
* txtmark 0.11 (https://github.com/rjeschke/txtmark)
* markdown4j 2.2 (https://code.google.com/p/markdown4j/)

## Benches

Benches are inspired from the [Henkelmann's Actuarius benches](http://henkelmann.eu/2011/01/10/performance_comparison_of_markdown_processor_for_the_jvm).

Benches are variations of the same text generate from `benches/plain.md`
This text is De finibus bonorum et malorum which is said to be the basis for
the commonly used Lorem Ipsum test text.

The generator rudely chops it up in a configurable number of paragraphs,
lines per paragraph and approximate chars per line.
Depending on the variation, it “decorates” parts of the text with Markdown syntax.
All tests consist of 30 Paragraphs with 20 lines each and approx. 80 chars per line (before “decoration”).

Here are the descriptions of the variations:

* Plain Paragraphs: No modifications, just the plain text.

* Emphasis: Every word emphasized,

  so the input `foo` is decorated as `*foo*` which should render as `<em>foo</em>`

* Strong: Every word emphasized,

  `foo` → `**foo**` → `<strong>foo</strong>`

* Inline Code: Every word marked as inline code:

  `foo` → ````foo```` → `<code>foo</code>`

* Fast Links: Every word a link without title or wrapped text:

  `foo` → `<foo>` → `<a href="foo">foo</a>`

* Special XML Chars: Every word is replaced by chars that need to be escaped in XML:

  `foo` → `"><&` → `&quot;&gt;&lt;'&amp;`

* Inline HTML: Every word is wrapped in verbatim HTML:

  `foo` → `<blink>foo</blink>` → `<blink>foo</blink>`

* Manual Line Breaks: Every line gets appended with `\n`:

  `some line` → `some line\n` → `some line<br/>\n`

* Full Links: Every word is turned into a link with wrapped text, url and title:

  `foo` → `[foo](http://example.com/foo "foo Title")` → `<a href="http://example.com/foo" title="foo Title">foo</a>`

* Full Images: Like full links, every word turned into an image reference

* Reference Links: Every word is turned into a link reference with an increasing id counter.
  The link reference definition is added after the paragraph the link occurs in:

  `foo` → `[foo][id123]` → `<a href="http://example.com/foo" title="foo Title">foo</a>`

* Block Quotes: Every paragraph is turned into a block quote by prepending a `>` to each line.

* Code Blocks: Every paragraph is turned into a code block by prepending four spaces to each line.

* Unordered Lists: Every paragraph is turned into a list with an item for each line by prepending `*` to each line.

* Mixed Test: A mix of all of the above: Some paragraphs lists, some code, some word emphasized etc.

Variation details can be found in `benches/gen_bench.nit`.

### CODE REVIEW

**Comment by Morriar** [comment]
And because you want numbers, here the result:


benches

Results for 50 consecutive runs.

Still some work to do on `nitmd` though.

**Comment by Morriar** [comment]
Those results are obtained for 1 run, before the JIT kicks in.

benches

More fair for `nitmd`, les for Java implems!

**Comment by privat** [comment]
C'est cool. 

Ca vaut le coup de valgrinder ou perfer les cas louches.

TODO: Bencher aussi l'original en perl et quelques uns en python et ruby pour que Nit ait moins l'air croche :)
Pourquoi pas aussi pandoc et discount (un écrit en C qui vise la perf)

File: benchmarks/markdown/benches/gen_benches.nit [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+		end
+		return res
+	end
+end
+
+# Wraps the word with `<blink>inline html<block>`.
+class InlineHtmlDecorator
+	super WordDecorator
+
+	redef fun decorate_word(w) do return "<blink>{w}</blink>"
```
Comment (privat) [inline benchmarks/markdown/benches/gen_benches.nit]: Lol blink

File: benchmarks/markdown/benches/gen_benches.nit [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+	print "gen_benches path/to/base/text.md"
+	exit 0
+end
+
+var out_dir = opt_dir.value or else "markdown.out/"
+out_dir.mkdir
+
+var file = new IFStream.open(ctx.rest.first)
+var txt = file.read_all
+file.close
```
Comment (privat) [inline benchmarks/markdown/benches/gen_benches.nit]: Et pourquoi pas l'excellente et moderne alternative

    var txt = ctx.rest.first.to_path.read_all

File: benchmarks/markdown/benches/plain.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+Quae fuerit causa, mox videro; interea hoc tenebo, si ob aliquam causam ista, quae sine dubio praeclara sunt, fecerint, virtutem iis per se ipsam causam non fuisse. -- Torquem detraxit hosti. -- Et quidem se texit, ne interiret. -- At magnum periculum adiit. -- In oculis quidem exercitus. -- Quid ex eo est consecutus? -- Laudem et caritatem, quae sunt vitae sine metu degendae praesidia firmissima. -- Filium morte multavit. -- Si sine causa, nollem me ab eo ortum, tam inportuno tamque crudeli; sin, ut dolore suo sanciret militaris imperii disciplinam exercitumque in gravissimo bello animadversionis metu contineret, saluti prospexit civium, qua intellegebat contineri suam. atque haec ratio late patet.
+In quo enim maxime consuevit iactare vestra se oratio, tua praesertim, qui studiose antiqua persequeris, claris et fortibus viris commemorandis eorumque factis non emolumento aliquo, sed ipsius honestatis decore laudandis, id totum evertitur eo delectu rerum, quem modo dixi, constituto, ut aut voluptates omittantur maiorum voluptatum adipiscendarum causa aut dolores suscipiantur maiorum dolorum effugiendorum gratia.
```
Comment (privat) [inline benchmarks/markdown/benches/plain.md]: honestatis toi-même!

File: benchmarks/markdown/engines/nitmd/nitmd.nit [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import markdown
+
+var file = args.first
+var n = args[1].to_i
+
+var ifs = new IFStream.open(file)
+var str = ifs.read_all
```
Comment (privat) [inline benchmarks/markdown/engines/nitmd/nitmd.nit]: Ferme ifs. Ou utilise to_path

**Comment by privat** [comment]
+1 mais c'est la dernière fois que je relis du pseudo latin

File: benchmarks/markdown/benches/gen_benches.nit [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+#
+# *This is not the markdown you are looking for.*
+interface MarkdownDecorator
+
+	# Parse the text and decorate it.
+	# Behavior depends on `LineDecorator` and `WordDecorator`.
+	fun decorate_text(txt: Text): Text is abstract
+
+	# Is `c` a something else than a letter a digit or a `-`?
+	fun is_workd_break(c: Char): Bool do
```
Comment (BlackMinou) [inline benchmarks/markdown/benches/gen_benches.nit]: workd ?

File: benchmarks/markdown/benches/gen_benches.nit [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+# Replaces the word with `[reflink][id123]`.
+class RefLinkDecorator
+	super WordDecorator
+
+	redef fun decorate_word(w) do
+		return "[{w}][id123]"
+	end
+end
+
+#
```
Comment (BlackMinou) [inline benchmarks/markdown/benches/gen_benches.nit]: "#" without comment ?

File: benchmarks/markdown/benches/gen_benches.nit [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+	private var line_decs: Array[LineDecorator] is lazy do
+		return [new ManualBreakDecorator, new BlockQuoteDecorator,
+		new CodeBlockDecorator, new UnorderedListDecorator: LineDecorator]
+	end
+
+	private var current_dec: LineDecorator = line_decs.first is lazy
+
+	redef fun decorate_line(line) do
+		# print line
```
Comment (BlackMinou) [inline benchmarks/markdown/benches/gen_benches.nit]: debug code ?

**Comment by Morriar** [comment]
Reroll:

* [x] Uses `path` interface
* [x] Adds comment for `MixedDecorator`
* [x] Removes debug code
* [x] Fixes typo `workd`

**Comment by Morriar** [comment]
More engines will come in further PRs.

**Comment by BlackMinou** [comment]
+1

File: benchmarks/markdown/benches/gen_benches.nit [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+# *This is not the markdown you are looking for.*
+interface MarkdownDecorator
+
+	# Parse the text and decorate it.
+	# Behavior depends on `LineDecorator` and `WordDecorator`.
+	fun decorate_text(txt: Text): Text is abstract
+
+	# Is `c` a something else than a letter a digit or a `-`?
+	fun is_word_break(c: Char): Bool do
+		if not c.is_letter or c.is_digit or c == '-' then
```
Comment (xymus) [inline benchmarks/markdown/benches/gen_benches.nit]: Doesn't `not is_letter` cover the other 2 cases? and this condition could be returned directly.

Comment (xymus) [inline benchmarks/markdown/benches/gen_benches.nit]: After reading the doc, shouldn't it be `not (c.is_letter or c.is_digit or c == '-')`?

File: benchmarks/markdown/engines/Makefile [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+all: nitmd/nitmd txtmark/Txtmark.class markdown4j/Markdown4j.class
+
+nitmd/nitmd:
+	cd nitmd; make
+
+txtmark/Txtmark.class:
+	cd txtmark; make
+
+markdown4j/Markdown4j.class:
+	cd markdown4j; make
```
Comment (xymus) [inline benchmarks/markdown/engines/Makefile]: You could use `make -C markdown4j`, it's cleaner in my opinion.

File: benchmarks/markdown/engines/txtmark/Makefile [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+txtmark-0.11.jar:
+	wget -O txtmark-0.11.jar http://search.maven.org/remotecontent?filepath=com/github/rjeschke/txtmark/0.11/txtmark-0.11.jar
+
+test: Txtmark.class
+	java -cp .:txtmark-0.11.jar Txtmark ../../benches/hello.md 5
+
+clean:
+	rm -rf txtmark-0.11.jar
+	rm -rf Txtmark.class
```
Comment (xymus) [inline benchmarks/markdown/engines/txtmark/Makefile]: No need for -r here.

**Comment by xymus** [comment]
+1, but the `if` is probably worth fixing.

**Comment by Morriar** [comment]
Reroll:

* [x] fixes `is_word_break`
* [x] uses `make -C` in Makefiles
* [x] removes useless `-rf`

**Pull request merged**